openapi3filter/validation_test.go

```diff
@@ -347,3 +354,369 @@ func toJSON(v interface{}) io.Reader {
+			},
+		},
+	}
+
+	// Create the test cases
+	tc := []struct {
+		name            string
+		schemes         *[]ExampleSecurityScheme
+		expectedSchemes *[]ExampleSecurityScheme
+	}{
+		{
+			name:    "/inherited-security",
+			schemes: nil,
+			expectedSchemes: &[]ExampleSecurityScheme{
+				securitySchemes[1],
+			},
+		},
+		{
+			name:            "/overwrite-without-security",
+			schemes:         &[]ExampleSecurityScheme{},
+			expectedSchemes: &[]ExampleSecurityScheme{},
+		},
+		{
+			name: "/overwrite-with-security",
+			schemes: &[]ExampleSecurityScheme{
+				securitySchemes[0],
+			},
+			expectedSchemes: &[]ExampleSecurityScheme{
+				securitySchemes[0],
+			},
+		},
+	}
+
+	// Create the swagger
+	swagger := &openapi3.Swagger{
+		Paths: map[string]*openapi3.PathItem{},
+		Security: openapi3.SecurityRequirements{
+			{
+				securitySchemes[1].Name: {},
+			},
+		},
+		Components: openapi3.Components{
+			SecuritySchemes: map[string]*openapi3.SecuritySchemeRef{},
+		},
+	}
+
+	// Add the security schemes to the components
+	for _, scheme := range securitySchemes {
+		swagger.Components.SecuritySchemes[scheme.Name] = &openapi3.SecuritySchemeRef{
+			Value: scheme.Scheme,
+		}
+	}
+
+	// Add the paths from the test cases to the swagger's paths
+	for _, tc := range tc {
+		var securityRequirements *openapi3.SecurityRequirements = nil
+		if tc.schemes != nil {
+			tempS := make(openapi3.SecurityRequirements, 0)
+			for _, scheme := range *tc.schemes {
+				tempS = append(
+					tempS,
+					openapi3.SecurityRequirement{
+						scheme.Name: {},
+					},
+				)
+			}
+			securityRequirements = &tempS
+		}
+		swagger.Paths[tc.name] = &openapi3.PathItem{
+			Get: &openapi3.Operation{
+				Security: securityRequirements,
+			},
+		}
+	}
+
+	// Declare the router
+	router := openapi3filter.NewRouter().WithSwagger(swagger)
+
+	// Test each case
+	for _, path := range tc {
+		// Make a map of the schemes and whether they're
+		var schemesValidated *map[*openapi3.SecurityScheme]bool = nil
+		if path.expectedSchemes != nil {
+			temp := make(map[*openapi3.SecurityScheme]bool)
+			schemesValidated = &temp
+			for _, scheme := range *path.expectedSchemes {
+				(*schemesValidated)[scheme.Scheme] = false
+			}
+		}
+
+		// Create the request
+		emptyBody := bytes.NewReader(make([]byte, 0))
+		pathUrl, err := url.Parse(path.name)
+		require.NoError(t, err)
+		route, _, err := router.FindRoute(http.MethodGet, pathUrl)
+		require.NoError(t, err)
+		req := openapi3filter.RequestValidationInput{
+			Request: httptest.NewRequest(http.MethodGet, path.name, emptyBody),
+			Route:   route,
+			Options: &openapi3filter.Options{
+				AuthenticationFunc: func(c context.Context, input *openapi3filter.AuthenticationInput) error {
+					if schemesValidated != nil {
+						if validated, ok := (*schemesValidated)[input.SecurityScheme]; ok {
+							if validated {
+								t.Fatalf("The path \"%s\" had the schemes %v named \"%s\" validated more than once",
```

Comment:
`%q` does exactly the same as `\"%s\"` for strings ;)

Reply:
I haven't taken properly learned go yet, so I didn't know that. Thanks a lot!